src/Native/Unix/System.Native/pal_process.c

```diff
@@ -144,6 +145,59 @@ static void ExitChild(int pipeToParent, int error)
     _exit(error != 0 ? error : EXIT_FAILURE);
 }
 
+static int SetGroups(uint32_t* userGroups, int32_t userGroupsLength, uint32_t* processGroups)
+{
+#ifdef __linux__
+    size_t platformGroupsLength = Int32ToSizeT(userGroupsLength);
+#else // BSD
+    int platformGroupsLength = userGroupsLength;
+#endif
+    int rv = setgroups(platformGroupsLength, userGroups);
+
+    // We fall back to using the current process' groups, if they are a subset of the user groups.
+    // We do this to support a user setting UserName to himself but not having setgroups permissions.
+    // And for dealing with platforms with low NGROUP_MAX (e.g. 16 on OSX).
+    if (rv == -1 && ((errno == EPERM) ||
+                     (errno == EINVAL && userGroupsLength > NGROUPS_MAX)))
+    {
+        int processGroupsLength = getgroups(userGroupsLength, processGroups);
+        if (processGroupsLength >= 0)
+        {
+            if (userGroupsLength == 0)
+            {
+                // calling setgroups with zero size returns number of groups.
+                rv = processGroupsLength == 0 ? 0 : -1;
+            }
+            else
+            {
+                rv = 0;
+                for (int i = 0; i < processGroupsLength; i++)
+                {
+                    bool isUserGroup = false;
+                    for (int j = 0; j < userGroupsLength && !isUserGroup; j++)
+                    {
+                        isUserGroup = processGroups[i] == userGroups[j];
+                    }
```

Comment:
Presumably the array lengths here are small enough that the O(processGroupsLength * userGroupsLength) doesn't have a negative perf impact?

Reply:
Yes. I'm making that assumption. If users are part of 10s of groups this is ok. If they are part of 100s then it would be better to optimize this. Should I improve this , e.g. qsort + bsearch?

Reply:
In our use case it is possible that a user belongs to 100+ groups. Especially when you're using an active directory backend where a user can belong to up to 1015, it would be better to optimize this.

Reply:
In our use case it is possible that a user belongs to 100+ groups. Especially when you're using an active directory backend where a user can belong to up to 1015, it would be better to optimize this.

Reply:
I've added the qsort + bsearch, this is now: O(processGroupsLength * log userGroupsLength)